### ISSUE TITLE
[BUG][Fuzzer][wrong-code] Block `T.tvm_thread_allreduce` over sub-4-byte dtypes silently miscomputes (codegen drops the warp barrier)

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2628]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11 (source pinned to tag `v0.1.11`, commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).

### System information

NVIDIA H100 80GB HBM3 (`sm_90`), Python 3.12, tilelang 0.1.11 (tag `v0.1.11` = `cd37ed5f`). The wrong values are **deterministic run-to-run** (20 launches of the same compiled kernel produce one identical wrong result), so this is a codegen defect (a missing barrier the compiler emitted then discarded), not a nondeterministic runtime race.

### Problem description

A block-level cross-thread reduction written with the low-level `T.tvm_thread_allreduce` intrinsic under a `reduce_scope` comm-reducer, over a **single warp**, silently produces the wrong result whenever the accumulator dtype is narrower than 4 bytes: `bfloat16`, `int16`, `int8`, `float8_*`. The identical reduction on `float16`, `float32`, or `int32` is correct.

For an 8-lane `sum` over inputs whose true total is `202` (`int16`/`int8`) or `41` (`bfloat16`):

| dtype | got | expected | correct? |
|---|---|---|---|
| `float32` (control) | `202` | `202` | ✓ |
| `int32` (control) | `202` | `202` | ✓ |
| `float16` (control) | `202` | `202` | ✓ |
| `bfloat16` | `23` | `41` | ✗ silent |
| `int16` | `90` | `202` | ✗ silent |
| `int8` | `90` | `202` | ✗ silent |

(int16/int8 rows use fixed inputs `[10,25,40,30,15,42,20,20]` summing to 202; bfloat16 uses `[3,8,5,2,7,6,4,6]` summing to 41, exact in bf16. The exact wrong value depends on the input ordering, but for a given input it is deterministic.) The wrong values are deterministic run-to-run (20 identical launches → one output). Onset is at the second reduction step: `reduce_threads=2` (one butterfly step) is correct; `4 / 8 / 16 / 32` are wrong. No error or warning is emitted.

The dtype boundary is **exactly** the `IsWarpReduction` filter (`ty.bytes() < 4`): the accepted dtypes take a `__shfl_down_sync` register path (correctly synchronized); the rejected sub-word dtypes fall to a shared-memory in-warp path whose intra-step barrier is silently discarded by codegen.

### Reproducible example code

```python
import torch
import tilelang as tl
import tilelang.language as T

def make_kernel(reduce_threads, dtype):
    @tl.jit(pass_configs={tl.PassConfigKey.TL_DISABLE_TMA_LOWER: True,
                          tl.PassConfigKey.TL_DISABLE_WARP_SPECIALIZED: True})
    def kernel(A):
        K = T.const("K"); A: T.Tensor((K,), dtype)
        C = T.empty((1,), dtype)
        with T.Kernel(1, threads=(reduce_threads,)):
            tk = T.get_thread_binding(0)          # reduce dim = threadIdx.x, single warp
            v = T.alloc_local((1,), dtype); v[0] = A[tk]
            Cr = T.alloc_local((1,), dtype)
            with T.attr(T.comm_reducer(lambda x, y: x + y, [T.cast(0, dtype)]),
                        "reduce_scope", T.reinterpret(T.uint64(0), dtype="handle")):
                T.evaluate(T.tvm_thread_allreduce(T.uint32(1), v[0], True, Cr[0], tk, dtype="handle"))
            C[0] = Cr[0]
        return C
    return kernel

# Fixed exact-integer inputs so the reported values reproduce byte-for-byte.
INT_VALS  = [10, 25, 40, 30, 15, 42, 20, 20]         # sum = 202
BF16_VALS = [3., 8., 5., 2., 7., 6., 4., 6.]         # sum = 41 (exact in bf16)
for dt, vals in [("float32", INT_VALS), ("int32", INT_VALS), ("float16", INT_VALS),
                 ("bfloat16", BF16_VALS), ("int16", INT_VALS), ("int8", INT_VALS)]:
    A = torch.tensor(vals, device="cuda").to(getattr(torch, dt))
    got = make_kernel(8, dt)(A).to(torch.float64)[0].item()
    ref = A.to(torch.float64).sum().item()
    print(f"{dt:9s} got={got}  ref={ref}  {'OK' if got == ref else 'WRONG (silent)'}")
# float32/int32/float16 -> OK ; bfloat16 got 23 (ref 41) ; int16/int8 got 90 (ref 202) -> WRONG (silent)
```

### Traceback

No traceback — the kernel compiles and runs; the result is silently wrong. (The inputs are exact integers, so any mismatch is a real defect, not float rounding.)

### Expected behavior

A block reduction must return the true reduction for every supported dtype, as `int32`/`float32`/`float16` already do. Either the sub-word in-warp path must be correctly synchronized (see fix), or the dtype should be rejected up front — not silently miscomputed.

### Additional context

**Root cause — codegen drops the `warp` barrier the pass relies on.**

Dtypes with `ty.bytes() < 4` are rejected by `IsWarpReduction` ([`lower_thread_allreduce.cc:850-878`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/lower_thread_allreduce.cc#L850-L878), the predicate `ty.bytes() < 4 || ty.bytes() > 8`), so they do **not** take the `__shfl_down_sync` register path. They fall to the shared-memory in-warp reduction `MakeBufAllreduce`, which brackets each butterfly step's load+store with `SyncThread("warp")` ([`lower_thread_allreduce.cc:774-778`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/lower_thread_allreduce.cc#L774-L778)):

```cpp
in_let_statement.emplace_back(SyncThread("warp"));   // barrier
in_let_statement.emplace_back(fstore(...));          // red_buf0[tx] = combine(...)
in_let_statement.emplace_back(SyncThread("warp"));   // barrier
```

`SyncThread("warp")` lowers to `tvm_storage_sync("warp")` ([`lower_thread_allreduce.cc:826-833`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/lower_thread_allreduce.cc#L826-L833)) — but the CUDA codegen lowers a `"warp"` storage-sync to **nothing**:

```cpp
// src/cuda/codegen/codegen_cuda.cc:1422-1424
if (sync == "warp") {
  // DO nothing.
}
```
([`codegen_cuda.cc:1422-1424`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/codegen/codegen_cuda.cc#L1422-L1424).) The pass wraps the three butterfly steps in a single `if (threadIdx.x < 4)` lane guard (the guard is correct — it keeps the 4 active lanes together) and relies on the two `SyncThread("warp")` barriers to order each step's load against the previous step's store. Because codegen discards those barriers, the emitted kernel has **no barrier between the steps inside the guard**, so the 4 active lanes read/write `red_buf0[tx]` and `red_buf0[tx+offset]` in the same shared buffer with no ordering:

```cuda
if (((int)threadIdx.x) < 4) {
  short w_4_0 = (red_buf0[((int)threadIdx.x)] + red_buf0[(((int)threadIdx.x) + 4)]);
  red_buf0[((int)threadIdx.x)] = w_4_0;
  short w_2_0 = (red_buf0[((int)threadIdx.x)] + red_buf0[(((int)threadIdx.x) + 2)]);  // no barrier between steps
  red_buf0[((int)threadIdx.x)] = w_2_0;                                               // reads a lane whose step-1 store may not be visible
  short w_1_0 = (red_buf0[((int)threadIdx.x)] + red_buf0[(((int)threadIdx.x) + 1)]);
  red_buf0[((int)threadIdx.x)] = w_1_0;
}
```

On Volta+ independent-thread scheduling nvcc is free to reorder these cross-lane reads/writes, yielding a deterministic wrong fold. The correct path (`int32`/`float32`/`float16`) instead emits `__shfl_down_sync(mask, red_buf0[0], offset, 32)`, which is implicitly synchronized — hence correct.

**Reachability — who hits this.** The path is reached only by writing the low-level `T.tvm_thread_allreduce` intrinsic under a `reduce_scope` comm-reducer with a **sub-4-byte accumulator dtype**. The four shipped examples that use this exact syntactic form — `examples/gemv/example_gemv.py`, `examples/dequantize_gemm/example_dequant_gemv_fp16xint4.py`, `examples/dequantize_gemm/example_dequant_gemm_fine_grained.py`, `examples/bitnet-1.58b/kernel_benchmark/tilelang_bitnet_158_int8xint2_decode.py` — all accumulate in `float32` (gemv default `accum_dtype=T.float`), `float16` (the dequant examples), or `int32` (bitnet, and the dp4a branch of the fp16xint4 dequant), i.e. the `IsWarpReduction`-accepted register-`__shfl` path (and fp16 has a dedicated correct lowering). **None of them accumulates in a sub-word dtype, so none triggers this bug** — a user only hits it by choosing a sub-word accumulator on the intrinsic. This is also **distinct from the public `T.reduce_sum`/`T.reduce_max` API**, which lowers through `src/op/reduce.cc` to the separate templated `tl::AllReduce<...>::run` in `reduce.h` (the #2542 family), never through `lower_thread_allreduce.cc`.

**Scope — the defect is op-agnostic and spans single- and multi-warp.** The dropped `SyncThread("warp")` bracket wraps the generic combiner `red_buf0[tx] = (*combiner)(red_buf0[tx], red_buf0[tx+off])` in `MakeBufAllreduce`, so it is independent of *which* reducer runs. Verified on H100 sm_90 (`int16` sub-word vs `float32` control, fixed inputs, deterministic ×3):

| reducer | single-warp (reduce=8) sub-word got / ref | multi-warp (reduce=64) sub-word got / ref |
|---|---|---|
| `sum` | 11 / 29 ✗ | int16 & int8 49 / 253 ✗ |
| `max` | −14 / 7 ✗ | 30 / 44 ✗ |
| `min` | −42 / −42 ✓ (extremum survived) | −42 / −45 ✗ |
| `prod` | 16 / 256 ✗ (only 4 of 8 factors fold) | — |

`fp32`/`int32`/`fp16` are exact for every reducer at both widths (control). The multi-warp branch still routes the intra-warp fold through the same `MakeBufAllreduce` (`IsWarpReduction` is false for sub-word), so it inherits the same dropped barrier. Two value-pattern caveats: (1) `min`/`max` are only *visibly* wrong when the extremum sits on a lane the corrupted fold drops — with a surviving extremum the result is accidentally correct (the `min` single-warp cell above); (2) `prod` needs a non-1 factor on the dropped lanes (all-1s reads correct). So `sum`/`prod` with uniform nonzero inputs make the defect visible most reliably. This *widens* the bug from "sub-word `sum`" to "sub-word block `tvm_thread_allreduce` for any reducer, single- or multi-warp" — one root, one fix; not separate bugs.

**Fix direction (probed on an isolated `.cu`).** A naive `__syncwarp()` placed *inside* the divergent step would deadlock (only the participating lanes reach it). Making codegen emit `__syncwarp` for `sync=="warp"` works only if the barrier is executed by all lanes (the pass already emits it outside a lane guard here, so a full-mask `__syncwarp()` at `codegen_cuda.cc:1422` is a candidate); alternatively, route the sub-word dtypes through the same mask-aware register-shuffle path the `≥4-byte` dtypes use, or reject `ty.bytes() < 4` at the frontend. The real fix touches the codegen no-op (the pass is emitting a barrier that codegen silently discards), so I'm phrasing it as options rather than a one-liner.

**Provenance.** The `"warp"` storage-sync no-op is from the initial code migration ([#10](https://github.com/tile-ai/tilelang/pull/10), `codegen_cuda.cc`); the shared-memory in-warp `SyncThread("warp")` bracket is from the upstream-TVM refactor ([#595](https://github.com/tile-ai/tilelang/pull/595), 2025-07-30). Both are long-standing — not a recent regression. No unit test exercises a sub-4-byte block `tvm_thread_allreduce`, so CI never caught it.

**Disposition.** wrong-code / silent-miscompute (say "silent" in the title): a valid kernel compiles and runs, returns wrong values, no diagnostic.

**Distinct from the existing reduce/sync issues:**
- **#2542** (`T.reduce_sum`/`T.reduce_max` non-power-of-two lane drop) — a *different template* (`tl::AllReduce` XOR-butterfly in `reduce.h`), triggered by non-pow2 thread counts. This bug fires at **power-of-two** counts (4/8/16/32) and is triggered by **dtype** (`bytes()<4`), on the `lower_thread_allreduce.cc` **pass** + codegen no-op.
- **#2599** (`T.warp_reduce_*` register builtin) — a different, user-facing warp-reduce builtin template.
- **#2556** (`tl.ThreadSync` drops `__syncthreads()` when the active count isn't a multiple of 32) — that pass drops a `shared` barrier on odd thread counts; here codegen drops a `warp` barrier that the AllReduce pass legitimately emits, at full-warp pow2 counts, gated on dtype.

I searched the tracker for `syncwarp`, `PrintStorageSync`, and `tvm_thread_allreduce bfloat16` and found no existing report.

<details>
<summary>Emitted CUDA (int16 buggy path vs float32 control)</summary>

`int16`, `reduce_threads=8` — verbatim `get_kernel_source()` (the shared-mem in-warp path; the three steps sit inside one lane guard with **no barrier between them**):
```cuda
v[0] = A[((int)threadIdx.x)];
__shared__ __align__(16) short red_buf0[8];
__syncthreads();
red_buf0[((int)threadIdx.x)] = v[0];
__syncthreads();
__syncthreads();
if (((int)threadIdx.x) < 4) {
  short w_4_0 = (red_buf0[((int)threadIdx.x)] + red_buf0[(((int)threadIdx.x) + 4)]);
  red_buf0[((int)threadIdx.x)] = w_4_0;
  short w_2_0 = (red_buf0[((int)threadIdx.x)] + red_buf0[(((int)threadIdx.x) + 2)]);
  red_buf0[((int)threadIdx.x)] = w_2_0;
  short w_1_0 = (red_buf0[((int)threadIdx.x)] + red_buf0[(((int)threadIdx.x) + 1)]);
  red_buf0[((int)threadIdx.x)] = w_1_0;
}
__syncthreads();
C[0] = red_buf0[0];
```
(The two `SyncThread("warp")` calls that should sit *between* the three steps inside the guard are what codegen dropped; the `__syncthreads()` are the `shared`-scope barriers around the buffer, which survive.) `float32`, `reduce_threads=8` — verbatim, the register-shuffle path (correct):
```cuda
float red_buf0[1];
red_buf0[0] = v[0];
mask[0] = __activemask();
t0[0] = __shfl_down_sync(mask[0], red_buf0[0], 4, 32); red_buf0[0] = (red_buf0[0] + t0[0]);
t0[0] = __shfl_down_sync(mask[0], red_buf0[0], 2, 32); red_buf0[0] = (red_buf0[0] + t0[0]);
t0[0] = __shfl_down_sync(mask[0], red_buf0[0], 1, 32); red_buf0[0] = (red_buf0[0] + t0[0]);
red_buf0[0] = __shfl_sync(mask[0], red_buf0[0], 0, 32);
C[0] = red_buf0[0];
```
</details>

**Comment by jjppp** [comment]
I think there should be a `__syncwarp()` when `sync == "warp"` in `codegen_cuda.cc`, or at least for GPUs newer than Volta.

After some digging, how tilelang handles warp synchronization seems to originate from some code written in 2017 in TVM, when we still don't have independent thread scheduling and `__syncwarp()`.

https://github.com/apache/tvm/commit/706f9b6f7eabf29aaa2f0d97c0d17fc92696c253#diff-74c87c7b8034b35a5f2c3e070cf6043d6528db645af6d1d872bc5e2c64d48aecR144-R146

**Comment by jjppp** [comment]
Some other notes on the test given in this issue: for `int8`, 202 should actually be `-54`, as its range falls within [-128, 127]
